### DISCUSSION TITLE
Fine-tuning Whisper in more than one language

### DISCUSSION THREAD

**Discussion by andrespimartin** [openai/whisper#1432] [Q&A]
Suppose I have a dataset in two or more languages (one of them under-represented in Whisper's pre-trained models), and I want to fine-tune those 2 or more languages to continue with a multilingual model and avoid catastrophic forgetting. Is fine-tuning possible?

Can I define the tokenizer and the processor without indicating the language?

**Comment by phineas-pta** [comment]
i don't think you can fine tune in one-click but you can do sequentially one language after another

**Reply by andrespimartin** [reply]
Hi @phineas-pta, thanks for your answer. 
In this way, would I not be creating a bias towards the last language to be trained?
Would I need to modify the learning rates as I train different languages?

**Reply by phineas-pta** [reply]
by bias, do you mean the model would misdetect a language by another?

**Reply by andrespimartin** [reply]
Yes, exactly. That and the model worsens its performance in those who are not the final language.

**Reply by phineas-pta** [reply]
sorry i have no idea, i think it shouldn't but idk any data to support my opinion

you should probably test it yourself: take any fine-tuned model on huggingface, evaluate the language identification performance and transcription performance on multiple languages

**Reply by andrespimartin** [reply]
hi, just to update.

I have been able to verify that with successive finetuning, catastrophic forgetting occurs, and the result gets worse on the first language.
However, I was able to launch the finetuning in a multilingual way, with more than one language at a time. I will update as soon as I have results.

**Reply by phineas-pta** [reply]
when you transcribe, do you let whisper auto detect or you tell exactly which language ?

**Reply by andrespimartin** [reply]
I am setting the language to the one corresponding to the test dataset

**Reply by ILG2021** [reply]
Hello, can you show me how do you do in the multilingual way? give the language to None?

**Reply by andrespimartin** [reply]
Hi @ILG2021 

You got two options:

1. Train without the language label and rely on the model capacity to detect the language.
2. Add the language label to label and train with this information. You can find more information on how to do this here: https://huggingface.co/spaces/openai/whisper/discussions/6#643d8bc551e2958ef6cd69ef

In my tests, both strategies show good performand on the ASR task. However, the second strategy also shows good results in the language identification task.

Hope this helps!

**Reply by ILG2021** [reply]
Thank you. I think the model will run faster when use method 2. this is my code below:

```
from dataclasses import dataclass
from typing import Any, List, Dict, Union
import re
import evaluate
import torch
from datasets import load_dataset, Audio, metric
from transformers import WhisperProcessor, WhisperForConditionalGeneration, \
    Seq2SeqTrainingArguments, Seq2SeqTrainer

model_name_or_path = 'openai/whisper-medium'
output_dir = "./whisper-medium-zh-tl"
data_dir = "./dataset"

processor = WhisperProcessor.from_pretrained(model_name_or_path, language=None, task="transcribe")

def prepare_dataset(batch):
    # load and resample audio data from 48 to 16kHz
    audio = batch["audio"]

    # compute log-Mel input features from input audio array
    batch["input_features"] = processor.feature_extractor(audio["array"], sampling_rate=audio["sampling_rate"]).input_features[0]
    processor.tokenizer.set_prefix_tokens(language=batch["language"], task="transcribe") 
    # encode target text to label ids
    batch["labels"] = processor.tokenizer(batch["transcription"]).input_ids
    return batch


@dataclass
class DataCollatorSpeechSeq2SeqWithPadding:
    processor: Any

    def __call__(self, features: List[Dict[str, Union[List[int], torch.Tensor]]]) -> Dict[str, torch.Tensor]:
        # split inputs and labels since they have to be of different lengths and need different padding methods
        # first treat the audio inputs by simply returning torch tensors
        input_features = [{"input_features": feature["input_features"]} for feature in features]
        batch = self.processor.feature_extractor.pad(input_features, return_tensors="pt")

        # get the tokenized label sequences
        label_features = [{"input_ids": feature["labels"]} for feature in features]
        # pad the labels to max length
        labels_batch = self.processor.tokenizer.pad(label_features, return_tensors="pt")

        # replace padding with -100 to ignore loss correctly
        labels = labels_batch["input_ids"].masked_fill(labels_batch.attention_mask.ne(1), -100)

        # if bos token is appended in previous tokenization step,
        # cut bos token here as it's append later anyways
        if (labels[:, 0] == self.processor.tokenizer.bos_token_id).all().cpu().item():
            labels = labels[:, 1:]

        batch["labels"] = labels
        return batch


def compute_metrics(pred):
    pred_ids = pred.predictions
    label_ids = pred.label_ids

    # replace -100 with the pad_token_id
    label_ids[label_ids == -100] = tokenizer.pad_token_id

    # we do not want to group tokens when computing the metrics
    pred_str = self.processor.tokenizer.batch_decode(pred_ids, skip_special_tokens=True)
    label_str = self.processor.tokenizer.batch_decode(label_ids, skip_special_tokens=True)

    wer = 100 * metric.compute(predictions=pred_str, references=label_str)

    return {"wer": wer}


if __name__ == "__main__":
    common_voice = load_dataset("audiofolder", data_dir=data_dir)
    common_voice = common_voice.cast_column("audio", Audio(sampling_rate=16000))
    common_voice = common_voice.map(prepare_dataset, remove_columns=common_voice.column_names["train"], num_proc=8)
    print(common_voice)
    train_samples = len(common_voice["train"])

    data_collator = DataCollatorSpeechSeq2SeqWithPadding(processor=processor)
    metric = evaluate.load("wer")
    model = WhisperForConditionalGeneration.from_pretrained(model_name_or_path)
    model.config.forced_decoder_ids = None
    model.config.suppress_tokens = []

    training_args = Seq2SeqTrainingArguments(
      output_dir=output_dir,  # change to a repo name of your choice
      per_device_train_batch_size=8,
      gradient_accumulation_steps=2,  # increase by 2x for every 2x decrease in batch size
      num_of_epchos=5,
      learning_rate=1e-5,
      warmup_steps=500,
      gradient_checkpointing=True,
      fp16=True,
      evaluation_strategy="steps",
      per_device_eval_batch_size=8,
      predict_with_generate=True,
      generation_max_length=225,
      save_steps=train_samples*10,   
      num_of_epchos/per_device_train_batch_size
      eval_steps=1000,
      logging_steps=25,
      report_to=["tensorboard"],
      load_best_model_at_end=True,
      metric_for_best_model="wer",
      greater_is_better=False,
      push_to_hub=False
    )

    trainer = Seq2SeqTrainer(
        args=training_args,
        model=model,
        train_dataset=common_voice["train"],
        eval_dataset=common_voice["test"],
        data_collator=data_collator,
        compute_metrics=compute_metrics,
        tokenizer=processor.feature_extractor,
    )

    trainer.train()
    trainer.save_model(output_dir)
```
I only change the prepare_dataset function. The batch size is 8. If a batch contains sample of 2 languages. is it ok? data_collator and compute_metrics will use tokenizer also. The load_dataset set language can also has an effect for these two?

**Reply by andrespimartin** [reply]
Hi @ILG2021 

Sorry for the delay, I've had a busy week.
Your code seems correct to me. You don't need to do anything else: since the language label is added, the model will try to predict that label as well, so it will be computed in the loss calculation. I have used datasets with six languages with the samples randomly concatenated, and a batch size of 16 and it works fine.

Best!

**Reply by ILG2021** [reply]
the self.processor.tokenizer.pad in DataCollatorSpeechSeq2SeqWithPadding is language independent? and self.processor.tokenizer.batch_decode don't need to set_prefix_tokens?

**Reply by tuandattt** [reply]
Hi, I use @ILG2021 code and just change my dataset but i still get this error:
ValueError: Multiple languages detected when trying to predict the most likely target language for transcription. It is currently not supported to transcribe to different languages in a single batch. Please make sure to either force a single language by passing language='...' or make sure all input audio is of the same language.

How can I solve this?

**Reply by sproocht** [reply]
Hi @tuandattt ,
Try transformers version 4.37.2 (or earlier). I opened an issue about this limitation yesterday: https://github.com/huggingface/transformers/issues/30654
All the best!

**Reply by sproocht** [reply]
Hi @tuandattt ,
This issue is now resolved in [this fix](https://github.com/huggingface/transformers/pull/30865) and merged into the latest version of transformers.

**Comment by andrespimartin** [comment]
As long as your language is included in the Whisper language, it will be correctly encoded and decoded, so yes, it is language independent.

Regarding the self.processor.tokenizer.batch_decode, it is used when computing the metrics for the ASR task, so it is correct to skip special tokens (you only want to compute the metric of the ASR task).
Note that you are including the language label in the loss calculation. If you want to compute the metrics for the LID task, you will need to override or rewrite the compute metrics function.

**Reply by ILG2021** [reply]
hello @andrespimartin , I tried to train with language label, but the evaluate wer is not that stable. When I train without language column, the wer is reduce stablely. But when I use the language label, it may increase much. Have you ever encountered this problem?

**Reply by andrespimartin** [reply]
Hi @ILG2021, I have been checking my training logs and yes, it seems that when you add the language label the eval WER does not reduce in a stable way. I haven't had time to look at the code, but I understand that the language label affects the way the WER is calculated, because in tests the results are good.

Best

**Reply by ILG2021** [reply]
When I finetune multi language, I found that if the languages is more than 3, it is very easy to overfitting. Every language's data is about 1.5 hours to 3 hours. Anyone encounted this problem?

**Comment by pr0mila** [comment]
I’ve created a code-switched language dataset for fine-tuning Whisper, including audio data along with CSV and Parquet files, which I’ve stored on Hugging Face. After preparing the dataset, I fine-tuned the model for translation. You can explore the entire end-to-end project in my repo. Here’s the link to check it out: https://github.com/pr0mila/MediBeng-Whisper-Tiny
